#! /bin/sh /usr/share/dpatch/dpatch-run
## 00_0183-qprocess-corruption.dpatch by Andreas Aardal Hanssen <ahanssen@trolltech.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix plain data loss bug.

@DPATCH@
diff -urNad qt4-x11-4.3.0~/src/corelib/io/qprocess.cpp qt4-x11-4.3.0/src/corelib/io/qprocess.cpp
--- qt4-x11-4.3.0~/src/corelib/io/qprocess.cpp	2007-05-25 15:24:09.000000000 +0200
+++ qt4-x11-4.3.0/src/corelib/io/qprocess.cpp	2007-06-30 19:53:44.000000000 +0200
@@ -826,15 +826,21 @@
     read(), readAll(), readLine(), and getChar(). It also determines
     which channel triggers QProcess to emit readyRead().
 
-    Changing the read channel will clear the unget buffer.
-
     \sa readChannel()
 */
 void QProcess::setReadChannel(ProcessChannel channel)
 {
     Q_D(QProcess);
-    if (d->processChannel != channel)
-        d->buffer.clear();
+    if (d->processChannel != channel) {
+        QByteArray buf = d->buffer.readAll();
+        if (d->processChannel == QProcess::StandardOutput) {
+            for (int i = buf.size() - 1; i >= 0; --i)
+                d->outputReadBuffer.ungetChar(buf.at(i));
+        } else {
+            for (int i = buf.size() - 1; i >= 0; --i)
+                d->errorReadBuffer.ungetChar(buf.at(i));
+        }
+    }
     d->processChannel = channel;
 }
 
